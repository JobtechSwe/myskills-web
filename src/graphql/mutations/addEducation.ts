import { InMemoryCache } from 'apollo-cache-inmemory'
import { Education } from '../../types'
import gql from 'graphql-tag'

export const GET_EDUCATIONS_CLIENT = gql`
  query getEducations {
    educations @client
  }
`

export const addEducation = (
  _: any,
  { education }: { education: Education },
  { cache }: { cache: InMemoryCache }
): Education => {
  const { educations }: any = cache.readQuery({
    query: GET_EDUCATIONS_CLIENT,
  })

  const withoutDuplicates = (educations: Education[]): Education[] =>
    educations.filter((e: Education) => e.taxonomyId !== education.taxonomyId)

  const updatedEducations = [...withoutDuplicates(educations), education]

  cache.writeQuery({
    query: GET_EDUCATIONS_CLIENT,
    data: { educations: updatedEducations },
  })

  return education
}
