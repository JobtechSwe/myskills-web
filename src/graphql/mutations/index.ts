export * from './addExperience'
export * from './addLanguage'
export * from './addSkill'
export * from './addEducation'
export * from './removeEducation'
export * from './removeExperience'
