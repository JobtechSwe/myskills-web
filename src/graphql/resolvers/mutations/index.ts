export * from './addLanguage'
export * from './addSkill'
export * from './addEducation'
export * from './addExperience'
export * from './addTrait'
export * from './addWhoAmI'
export * from './createOccupation'
export * from './updateContactInformation'
export * from './removeEducation'
export * from './removeExperience'
export * from './removeOccupation'
export * from './removeSkill'
export * from './removeTrait'
export * from './updateEducation'
export * from './updateExperience'
export * from './uploadImage'
