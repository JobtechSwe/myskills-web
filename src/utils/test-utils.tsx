import * as React from 'react'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo-hooks'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink, MockedProvider } from 'react-apollo/test-utils'
import { ThemeProvider } from 'emotion-theming'
import { render as rtlRender } from 'react-testing-library'
import { theme } from 'theme'

export const render = (children: React.ReactNode, mocks?: any[]) => {
  const client = new ApolloClient({
    cache: new InMemoryCache({ addTypename: true }),
    link: new MockLink(mocks || []),
  })

  return rtlRender(
    <MockedProvider addTypename={true} mocks={mocks}>
      <ApolloProvider client={client}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </ApolloProvider>
    </MockedProvider>
  )
}
