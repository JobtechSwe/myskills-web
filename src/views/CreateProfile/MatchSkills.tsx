import ButtonToInput from 'components/ButtonToInput'
import Flex from 'components/Flex'
import React from 'react'
import TagList from 'components/TagList'
import { H1 } from 'components/Typography'
import { RouteComponentProps } from '@reach/router'
import { useQuery, useMutation } from 'react-apollo-hooks'
import { v4 } from 'uuid'
import { withApollo, WithApolloClient } from 'react-apollo'
import { ADD_SKILL_CLIENT, REMOVE_SKILL_CLIENT } from 'graphql/shared/Mutations'
import {
  GET_RELATED_SKILLS,
  GET_SKILLS_CLIENT,
  GET_OCCUPATION_CLIENT,
} from 'graphql/shared/Queries'
import {
  OntologyType,
  OntologyRelationResponse,
  Skill,
  SkillInput,
} from 'generated/myskills.d'
import RegistrationLayout from 'components/Layout/RegistrationLayout'

const MatchSkills: React.FC<WithApolloClient<RouteComponentProps>> = ({
  client,
}) => {
  const {
    data: { occupation = {} },
  }: any = useQuery(GET_OCCUPATION_CLIENT)

  const {
    data: { skills = [] },
  }: any = useQuery(GET_SKILLS_CLIENT)

  const addSkillMutation = useMutation(ADD_SKILL_CLIENT)
  const removeSkillMutation = useMutation(REMOVE_SKILL_CLIENT)

  const [relatedSkills, setRelatedSkills] = React.useState<
    OntologyRelationResponse[]
  >([])

  const getRelatedSkills = React.useCallback(
    async (skills: any[]) => {
      const { data } = await client.query({
        query: GET_RELATED_SKILLS,
        variables: {
          concepts: skills.map(({ term }) => term),
          limit: 5,
          type: OntologyType.Skill,
        },
      })

      setRelatedSkills(data.ontologyRelated.relations)
    },
    [client]
  )

  React.useEffect(() => {
    skills.length ? getRelatedSkills(skills) : getRelatedSkills([occupation])
  }, [skills, occupation, getRelatedSkills])

  const handleSkillClick = (skill: Skill) => {
    const hasSkill = skills.some((s: Skill) => s.term === skill.term)

    if (hasSkill) {
      removeSkillMutation({
        variables: {
          skill,
        },
      })
    } else {
      addSkillMutation({
        variables: {
          skill,
        },
      })
    }
  }

  const handleFreeTextSkill = (value: string) => {
    const skill = {
      term: value,
      type: OntologyType.Skill,
      sourceId: v4(),
    }

    addSkillMutation({
      variables: {
        skill,
      },
    })
  }

  const ontologyRelationToSkill = (
    ontologyItem: OntologyRelationResponse
  ): SkillInput => ({
    sourceId: ontologyItem.id,
    term: ontologyItem.term,
    type: ontologyItem.type,
  })

  return (
    <RegistrationLayout headerText="KOMPETENS" nextPath="erfarenheter" step={2}>
      <Flex
        alignItems="center"
        flexDirection="column"
        justifyContent="flex-start"
      >
        <H1 mb={20}>Vilka är dina kompetenser?</H1>
        <TagList
          activeItems={skills.map((s: SkillInput) => ({
            ...s,
            id: s.sourceId,
          }))}
          items={relatedSkills
            .map(ontologyRelationToSkill)
            .filter(
              (relatedSkill: SkillInput) =>
                !skills.some((s: SkillInput) => s.term === relatedSkill.term)
            )
            .map(s => ({ ...s, id: s.sourceId }))}
          onSelect={handleSkillClick}
        />
        <ButtonToInput
          buttonText="+ Lägg till en kompetens"
          inputPlaceholder="Lägg till en kompetens"
          onSelect={handleFreeTextSkill}
        />
      </Flex>
    </RegistrationLayout>
  )
}

export default withApollo(MatchSkills)
