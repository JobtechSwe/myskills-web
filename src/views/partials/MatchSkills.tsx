import ButtonToInput from 'components/ButtonToInput'
import Flex from 'components/Flex'
import React from 'react'
import TagList from 'components/TagList'
import { H1, Paragraph } from 'components/Typography'
import { RouteComponentProps } from '@reach/router'
import { MutationHookOptions } from 'react-apollo-hooks'
import { v4 } from 'uuid'
import { withApollo, WithApolloClient } from 'react-apollo'
import { GET_RELATED_SKILLS } from 'graphql/shared/Queries'
import {
  Occupation,
  OntologyRelationResponse,
  OntologyType,
  Skill,
  MutationRemoveSkillArgs,
  MutationAddSkillArgs,
} from 'generated/myskills.d'
import { FooterButton } from 'components/Layout/Registration'

interface ISkillProps {
  term: string
}

interface MatchSkillsProps {
  buttonText: string
  onSubmit: () => void
  addSkill: (skill: MutationHookOptions<{}, MutationAddSkillArgs>) => any
  removeSkill: (id: MutationHookOptions<{}, MutationRemoveSkillArgs>) => any
  occupation: Occupation
  skills: Skill[]
}

const MatchSkills: React.FC<
  WithApolloClient<RouteComponentProps & MatchSkillsProps>
> = ({
  buttonText,
  client,
  onSubmit,
  skills,
  occupation,
  addSkill,
  removeSkill,
}) => {
  const [relatedSkills, setRelatedSkills] = React.useState<
    OntologyRelationResponse[]
  >([])

  const getRelatedSkills = React.useCallback(
    async (terms: (Skill | Occupation)[]) => {
      const { data } = await client.query({
        query: GET_RELATED_SKILLS,
        variables: {
          concepts: terms.map(({ term }) => term),
          limit: 5,
          type: OntologyType.Skill,
        },
      })

      setRelatedSkills(data.ontologyRelated.relations)
    },
    [client]
  )

  React.useEffect(() => {
    skills.length ? getRelatedSkills(skills) : getRelatedSkills([occupation])
  }, [skills, occupation, getRelatedSkills])

  const handleSkillClick = (skill: Skill) => {
    const hasSkill = skills.some(s => s.term === skill.term)
    if (hasSkill) {
      removeSkill({
        variables: {
          id: skill.id,
        },
      })
    } else {
      addSkill({
        variables: {
          skill,
        },
      })
    }
  }

  const handleFreeTextSkill = (value: string) => {
    const skill = {
      term: value,
      type: OntologyType.Skill,
      sourceId: v4(),
      id: v4(),
    }

    addSkill({
      variables: {
        skill,
      },
    })
  }

  const ontologyRelationToSkill = (
    ontologyItem: OntologyRelationResponse
  ): Skill => ({
    sourceId: ontologyItem.id,
    term: ontologyItem.term,
    type: ontologyItem.type,
    id: ontologyItem.id,
  })

  return (
    <>
      <Flex
        alignItems="center"
        flexDirection="column"
        justifyContent="flex-start"
      >
        <H1 mb={20}>Vilka är dina kompetenser?</H1>
        <Paragraph mb="large" mt={0} textAlign="center">
          Ditt val av yrkesroll matchar ofta nedan föreslagna kompetenser.
          Markera dina eller lägg till egna.
        </Paragraph>
        <TagList
          activeItems={skills}
          items={relatedSkills
            .map(ontologyRelationToSkill)
            .filter(
              relatedSkill => !skills.some(s => s.term === relatedSkill.term)
            )}
          onSelect={handleSkillClick}
        />
        <ButtonToInput
          buttonText="+ Lägg till kompetens"
          inputPlaceholder="Lägg till kompetens"
          onSelect={handleFreeTextSkill}
        />
      </Flex>
      <FooterButton onClick={onSubmit} text={buttonText} />
    </>
  )
}

export default withApollo(MatchSkills)
