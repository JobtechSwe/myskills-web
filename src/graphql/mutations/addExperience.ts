import { InMemoryCache } from 'apollo-cache-inmemory'
import { Experience } from '../../types'
import { storageHelper } from '../../utils/helpers'
import gql from 'graphql-tag'

export const GET_EXPERIENCES_CLIENT = gql`
  query getExperiences {
    experiences @client {
      name
      taxonomyId
      years
    }
  }
`

export const addExperience = (
  _: any,
  { experience }: { experience: Experience },
  { cache }: { cache: InMemoryCache }
): Experience => {
  const { experiences }: any = cache.readQuery({
    query: GET_EXPERIENCES_CLIENT,
  })

  const withoutDuplicates = (exp: Experience[]): Experience[] =>
    exp.filter((e: Experience) => e.taxonomyId !== experience.taxonomyId)

  const updatedExperiences = [...withoutDuplicates(experiences), experience]

  cache.writeQuery({
    query: GET_EXPERIENCES_CLIENT,
    data: { experiences: updatedExperiences },
  })

  storageHelper.set({
    type: 'experiences',
    data: updatedExperiences,
  })

  return experience
}
