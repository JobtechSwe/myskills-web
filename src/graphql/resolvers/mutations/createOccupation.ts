import { InMemoryCache } from 'apollo-cache-inmemory'
import { storageHelper } from '../../../utils/helpers'
import gql from 'graphql-tag'

export const GET_OCCUPATION_CLIENT = gql`
  query occupation {
    occupation @client {
      term
      experience {
        years
      }
    }
  }
`

export const createOccupationClient = (
  _: any,
  { occupation }: any,
  { cache }: { cache: InMemoryCache }
): any => {
  cache.writeQuery({
    query: GET_OCCUPATION_CLIENT,
    data: {
      occupation,
    },
  })

  storageHelper.set({
    type: 'occupation',
    data: occupation,
  })

  return occupation
}
