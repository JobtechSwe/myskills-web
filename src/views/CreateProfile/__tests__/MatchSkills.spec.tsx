import * as React from 'react'
import MatchSkills, { GET_RELATED_SKILLS } from '../MatchSkills'
import { render } from '../../../utils/test-utils'
import { act, wait } from 'react-testing-library'
import { OntologyType } from '../../../generated/myskills.d'
import { GET_OCCUPATION_CLIENT } from '../../../graphql/resolvers/mutations/createOccupation'

describe('views/MatchSkills', () => {
  it('renders loading state', async done => {
    const withResultsMock = [
      {
        request: {
          query: GET_RELATED_SKILLS,
          variables: {
            concepts: [],
            type: OntologyType.Skill,
            limit: 5,
          },
        },
        result: {
          data: {
            ontologyRelated: {
              relations: [],
              __typename: 'asdf',
            },
          },
        },
      },
    ]
    const { container } = render(<MatchSkills />, withResultsMock)
    act(() => {
      expect(container).toMatchSnapshot()
      done()
    })
  })

  it('should render with ontology related query result', () => {
    const withResultsMock = [
      {
        request: {
          query: GET_OCCUPATION_CLIENT,
        },
        result: {
          data: {
            occupation: {
              name: 'snickare',
              experience: {
                years: 5,
              },
              __typename: 'asdf',
            },
          },
        },
      },
      {
        request: {
          query: GET_RELATED_SKILLS,
          variables: {
            concepts: ['snickare'],
            limit: 5,
            type: OntologyType.Skill,
          },
        },
        result: {
          data: {
            ontologyRelated: {
              relations: [
                {
                  name: 'Händig',
                  score: 0.5,
                  id: '123',
                  type: OntologyType.Skill,
                  __typename: 'assdf',
                },
              ],
              __typename: 'asdf',
            },
          },
        },
      },
      {
        request: {
          query: GET_RELATED_SKILLS,
          variables: {
            concepts: [],
            limit: 5,
            type: OntologyType.Skill,
          },
        },
        result: {
          data: {
            ontologyRelated: {
              relations: [],
              __typename: 'asdf',
            },
          },
        },
      },
    ]

    const { getByText } = render(<MatchSkills />, withResultsMock)
    wait().then(() => {
      expect(getByText(/händig/i)).toBeInTheDocument()
    })
  })
})
