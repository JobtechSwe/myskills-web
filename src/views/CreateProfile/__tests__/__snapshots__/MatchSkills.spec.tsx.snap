// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/MatchSkills renders loading state 1`] = `
<div>
  <div
    class="css-15q7p9b"
    height="calc(var(--vh, 1vh) * 100)"
  >
    <nav
      class="css-208zbf"
    >
      <div
        class="css-nshvkv"
      >
        <div
          class="css-cn959j"
          role="button"
          tabindex="0"
        >
          <img
            alt="Go back"
            src="navigation_arrow.svg"
          />
        </div>
        <div
          class="css-1a620cz"
        >
          <div
            class="css-njydrf"
          />
          <div
            class="css-njydrf"
          />
          <div
            class="css-1fl5aa0"
          />
          <div
            class="css-1fl5aa0"
          />
          <div
            class="css-1fl5aa0"
          />
          <div
            class="css-1fl5aa0"
          />
        </div>
      </div>
      <p
        class="css-1wtsfx0"
        font-family="default"
        font-size="small"
      >
        KOMPETENS
      </p>
    </nav>
    <div
      class="css-13lzt9w"
    >
      
      <h1
        class="css-1h19gqr"
        color="black"
        font-family="default"
        font-size="large"
        font-weight="600"
      >
        Vilka är dina kompetenser?
      </h1>
      <span
        class="css-bhiavv"
        cursor="pointer"
        opacity="1"
        role="button"
        tabindex="0"
      >
        + Lägg till en kompetens
      </span>
      <div
        class="sk-fading-circle css-rcik79"
      >
        <div
          class="sk-circle1 sk-circle"
        />
        <div
          class="sk-circle2 sk-circle"
        />
        <div
          class="sk-circle3 sk-circle"
        />
        <div
          class="sk-circle4 sk-circle"
        />
        <div
          class="sk-circle5 sk-circle"
        />
        <div
          class="sk-circle6 sk-circle"
        />
        <div
          class="sk-circle7 sk-circle"
        />
        <div
          class="sk-circle8 sk-circle"
        />
        <div
          class="sk-circle9 sk-circle"
        />
        <div
          class="sk-circle10 sk-circle"
        />
        <div
          class="sk-circle11 sk-circle"
        />
        <div
          class="sk-circle12 sk-circle"
        />
      </div>
    </div>
    <div
      class="css-djo2sl"
    >
      <button
        class="css-1fyzulv"
        data-testid="okButton"
        font-family="default"
        font-size="medium"
      >
        Fortsätt
      </button>
    </div>
  </div>
</div>
`;
